### PR TITLE
[build][android] set INSTALL_RPATH properly for shared libraries

### PR DESCRIPTION
**Pull request by finagolfin**
~This finally allows removing the toolchain/SDK rpaths that are currently automatically added when compiling Swift shared libraries and executables on non-Darwin platforms. Also,~ Similar flag added in #30430 instead. Remove unnecessary checks for Android's INSTALL_RPATH, now that the CMake setup has been refactored to separate host and target library configuration.

~I've [backported this swiftc flag to the Swift 5.1.4 package script for Android](https://github.com/buttaface/termux-packages/commit/8e733afe84b7a8d0324a4825abe6e610ea3a17aa#diff-f2454b0e4d1c9024e7f8ba47e3d7703aR20) and then use it [to avoid adding the toolchain rpath for Foundation](https://github.com/buttaface/termux-packages/commit/8e733afe84b7a8d0324a4825abe6e610ea3a17aa#diff-3cdf9fd33dc2ba4aa8334a1067babff2R19) and other Swift shared libraries in the toolchain.~

~That would be useful for the official releases for linux too, here's what the 5.1.4 release and the latest 5.2 snapshot show:~
```
> readelf -d swift-5.1.4-RELEASE-ubuntu18.04/usr/lib/swift/linux/libFoundation.so | ag runpath
0x000000000000001d (RUNPATH)            Library runpath: [/home/buildnode/jenkins/workspace/oss-swift-5.1-package-linux-ubuntu-18_04/build/buildbot_linux/swift-linux-x86_64/lib/swift/linux:/home/buildnode/jenkins/workspace/oss-swift-5.1-package-linux-ubuntu-18_04/build/buildbot_linux/libdispatch-linux-x86_64/src:$ORIGIN]
> readelf -d swift-5.2-DEVELOPMENT-SNAPSHOT-2020-03-03-a-ubuntu18.04/usr/lib/swift/linux/libFoundation.so | ag runpath
0x000000000000001d (RUNPATH)            Library runpath: [/home/buildnode/jenkins/workspace/oss-swift-5.2-package-linux-ubuntu-18_04/build/buildbot_linux/swift-linux-x86_64/lib/swift/linux:/home/buildnode/jenkins/workspace/oss-swift-5.2-package-linux-ubuntu-18_04/build/buildbot_linux/libdispatch-linux-x86_64:$ORIGIN]
```
~@brentdax, you modified some of this rpath config last year, let me know what you think.~ @drodriguez, let me know about the Android part. ~@tachoknight, this should be useful for all Swift packagers.~

### CODE REVIEW

File: cmake/modules/AddSwift.cmake [hunk 1]
```diff
@@ -628,14 +628,9 @@ function(_add_swift_host_library_single target)
...
-    # case both the NDK and Sysroot paths must be set.
-    if(NOT "${SWIFT_ANDROID_NDK_PATH}" STREQUAL "" AND
-       NOT "${SWIFT_ANDROID_NATIVE_SYSROOT}" STREQUAL "")
-      set_target_properties("${target}"
-        PROPERTIES
-        INSTALL_RPATH "$ORIGIN")
-    endif()
+    set_target_properties("${target}"
+      PROPERTIES
+      INSTALL_RPATH "$ORIGIN")
```
Comment (buttaface) [inline cmake/modules/AddSwift.cmake]: No checks needed now that this is only set for host libraries on Android.

**Comment by buttaface** [comment]
Ping, this is a small pull that would help get rid of unnecessary rpaths in Swift binaries and shared libraries on ELF platforms.

**Comment by drodriguez** [comment]
I don't know about the logic of the changes (if they are right or not), but it doesn't break the Android builds for me, so I hope it doesn't break them in CI.

But this changes the Linux builds slightly, in a way similar to what the `DarwinToolchain.cpp` seems to do, so it will affect more than Android. If anyone has any objections, or worries, it will be nice to know them before this moves further.

@swift-ci please test

**Comment by buttaface** [comment]
All this pull does is allow those on ELF platforms, ie linux, Android, BSD, and so on, to disable the host toolchain rpath from being automatically added to Swift shared libraries by manually adding the `-no-stdlib-rpath` flag, [which is already available and working on macOS](https://github.com/apple/swift/blob/655d89b14610ec5628ed14922a0e70ee9c43d3fb/lib/Driver/DarwinToolChains.cpp#L412). Given the example in the OP, where the build/CI toolchain rpath has been wrongly added to the official release of the Foundation shared library for linux and now is invoked on any other host on which that official linux toolchain Foundation is run, this flag would optionally allow disabling adding that toolchain rpath, as I show in the OP for Foundation with my linked Android Swift 5.1.4 packaging script patch.

Since this merely enables an optional flag that is already defined for ELF platforms too, it should have no effect on the pre-existing tests, builds, and so on. It merely gives those who know what they're doing a flag they can use to disable that toolchain rpath from being added to Swift ELF shared libraries, just like on macOS.

**Comment by buttaface** [comment]
Oh, the macOS CI failure seems spurious.

**Comment by buttaface** [comment]
Saleem just merged #30430 which adds `-no-toolchain-stdlib-rpath` instead, so I removed the `-no-stdlib-rpath` commit in this pull. All that's left is the simple changes to the `INSTALL_RPATH`.

I haven't added the native sysroot location, ie something like `/usr/lib/swift/linux` above instead of just `$ORIGIN` alone, to the install rpath, as I believe that's a bad idea and actually shouldn't be done for linux either. It is very likely that the versions of the different target libraries from the separate rpaths wouldn't be the same and it would cause more confusion than help. I can remove those for linux/cygwin too in this pull, if the reviewers agree.

**Comment by drodriguez** [comment]
@swift-ci please test and merge

**Comment by buttaface** [comment]
Hmm, it didn't go through, but that's probably for the best, as I think I'll have to change this pull a bit. Hold off on merging for now, I'll update this pull in a day or two.

**Comment by buttaface** [comment]
I don't think I'll add anything else after all, please go ahead and merge.

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by buttaface** [comment]
Linux failure is spurious.

**Comment by drexin** [comment]
@swift-ci smoke test linux

**Comment by buttaface** [comment]
Ready to go? This only affects Android, so very safe to merge.

**Pull request merged**